core/arch/aarchxx/mangle.c

```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
 }
 #endif /* !AARCH64 */
 
+/* Return true if opnd is a register, but not XSP, or immediate zero on AArch64. */
+static bool
+opnd_is_reglike(opnd_t opnd)
+{
+    return ((opnd_is_reg(opnd) && opnd_get_reg(opnd) != DR_REG_XSP) ||
+            IF_X64_ELSE((opnd_is_immed_int(opnd) && opnd_get_immed_int(opnd) == 0),
+                        false));
+}
+
 uint
 insert_parameter_preparation(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
                              bool clean_call, uint num_args, opnd_t *args)
 {
-    uint i;
-    instr_t *mark = INSTR_CREATE_label(dcontext);
-    PRE(ilist, instr, mark);
+    uint num_regs = num_args < NUM_REGPARM ? num_args : NUM_REGPARM;
+    signed char regs[NUM_REGPARM];
+    int usecount[NUM_REGPARM];
+    ptr_int_t stack_inc = 0;
+    uint i, j;
 
-    ASSERT(num_args == 0 || args != NULL);
-    /* FIXME i#1551, i#1569: we only support limited number of args for now. */
-    ASSERT_NOT_IMPLEMENTED(num_args <= NUM_REGPARM);
+    /* We expect every arg to be an integer or a full-size register. */
```

Comment:
I'm slightly unsure about which memory operands should be accepted:

constant offset from a register: Yes.
register offset from a register: A pain to implement but hard to justify excluding.
REL_ADDR_kind: Is this useful? (It would be AArch32-only.)
MEM_INSTR_kind: Does that make any sense here?
ABS_ADDR_kind: Does not exist for ARM/AArch64.